### DISCUSSION TITLE
A Big Hi, Hello 👋: API & Webhooks Community Check-In

### DISCUSSION THREAD

**Discussion by ghostinhershell** [community/community#31405] [Apps, API and Webhooks]
## In a nutshell 🐿️
What's the difference between REST API and GraphQL, and is one query language more efficient than the other?

--

Hey guys, [Sarto](https://github.com/ghostinhershell) here 👋 I’m the new Technical Community Manager over in the API and Webhooks area of our community (2 months in, woohoo! 🎉).

When looking through the questions that are commonly asked in the API and Integrations area of our community, I noticed that questions related to GraphQL were frequently coming up. Although I’m familiar with [REST APIs](https://docs.github.com/en/rest), GraphQL is a query language that I’m not very familiar with. So, I set out to answer one simple question: **What’s the difference between REST API and GraphQL, and is one query language more efficient than the other?**

## Key Findings - The Meat & Potatoes 🍗🥔
For the fans of using REST, you can _rest_ assured that this isn’t a post that plans on convincing you otherwise. However, there are a couple of things I’ve learned that makes switching from REST to GraphQL pretty compelling:

- Declare what you want, get exactly what you need. You don’t have to worry about fetching more or less data than what you’re looking for. Ultimately, it can be more performant. 
- [GraphQL uses a single endpoint](https://docs.github.com/en/graphql/guides/forming-calls-with-graphql#the-graphql-endpoint). That means that the endpoint remains consistent no matter what operation you perform.
- You can nest requests using GraphQL, cutting down on the number of requests you would need to make if you were using REST.

I think those are intriguing enough to teach a dog new tricks.

One thing worth noting, for me at least, is that there seems to be a bit of a learning curve for GraphQL. For example, it took skimming through a couple of YouTube videos and reading through our support docs in order to write my first successful query. I’m more of a hands-on learner so I think that with practice, I’ll be more comfortable with the language.

## Helpful Resources 📚
Here are some helpful GitHub Docs:

- [Introduction to GraphQL](https://docs.github.com/en/graphql/guides/introduction-to-graphql)
- [Getting Started with REST API](https://docs.github.com/en/rest/guides/getting-started-with-the-rest-api)
- [Migrating from REST to GraphQL](https://docs.github.com/en/graphql/guides/migrating-from-rest-to-graphql)

Here’s a few external resources I would recommend checking out as well:

- [Intro to GraphQL](https://graphql.org/learn/)
- [A Beginners Guide to GraphQL](https://www.freecodecamp.org/news/a-beginners-guide-to-graphql-86f849ce1bec/)
- [GraphQL Explained in 100 Seconds](https://www.youtube.com/watch?v=eIQh02xuVw4)
- Trulia uses GraphQL as part of their architecture: [One Endpoint to Rule Them All](https://www.trulia.com/blog/tech/graphql-one-endpoint-to-rule-them-all/)

**Have any other helpful resources? Please share in the comments!**

Looking forward to hearing from you folks, what’s your favorite query language? Is there a reason you prefer to use one over the other? Got any fun facts or tips to share with us GraphQL newbs? Sound off below 😀

Cheers!

**Comment by naytun** [comment]
GraphQL is an **interface layer** between the database and front-end. Also, GraphQL enforces **type** to the fields used in the query. Similar to TypeScript is to Javascript.

**Reply by ghostinhershell** [reply]
> Similar to TypeScript is to Javascript.

I think this is a great comparison to make, thanks so much for chiming in here @naytun! 😄
